### PR TITLE
seamless background pattern

### PR DESCRIPTION
**Pull request by mollymerp**
similar implementation of #1908 but for background-pattern, will solve #1914 

the seamless background pattern is working at this stage, but I'm not sure how to go about testing it.

what I have done so far has introduced a bug with the map rendering after the initial load, so I will be working on tracking that down as well.

### CODE REVIEW

File: debug/site.js [hunk 1]
```diff
@@ -44,6 +44,16 @@ map.on('load', function() {
...
+        "type":"background",
+        "source":"composite",
+        "paint": {
+            "background-pattern":"fast-food-15"
+        }
+    });
+
     map.addLayer({
         "id": "random-points",
         "type": "circle",
```
Comment (lucaswoj) [inline debug/site.js]: Please revert your changes to this file.

File: js/render/draw_background.js [hunk 2]
```diff
@@ -3,7 +3,7 @@
 var TilePyramid = require('../source/tile_pyramid');
 var pyramid = new TilePyramid({ tileSize: 512 });
 var util = require('../util/util');
-var EXTENT = require('../data/bucket').EXTENT;
+var extent = 8192;
```
Comment (lucaswoj) [inline js/render/draw_background.js]: Why did you change this line? It is important that we keep a single canonical definition of `EXTENT`.

File: js/render/draw_background.js [hunk 3]
```diff
@@ -70,10 +58,48 @@ function drawBackground(painter, source, layer) {
...
         gl.drawArrays(gl.TRIANGLE_STRIP, 0, painter.tileExtentBuffer.itemCount);
     }
 
     gl.stencilMask(0x00);
     gl.stencilFunc(gl.EQUAL, 0x80, 0x80);
 }
+
+function pixelsToTileUnits(coord, pixelValue, z) {
+    return pixelValue * (extent / (512 * Math.pow(2, z - coord.z)));
+}
```
Comment (lucaswoj) [inline js/render/draw_background.js]: Can you find a way to eliminate the code duplication between :point_up_2: and [`Tile`](https://github.com/mapbox/mapbox-gl-js/blob/master/js/source/tile.js#L43)?

File: js/render/draw_fill.js [hunk 4]
```diff
@@ -12,7 +12,6 @@ function draw(painter, source, layer, coords) {
     var color = util.premultiply(layer.paint['fill-color'], layer.paint['fill-opacity']);
     var image = layer.paint['fill-pattern'];
     var strokeColor = util.premultiply(layer.paint['fill-outline-color'], layer.paint['fill-opacity']);
-
     // Draw fill
     if (image ? !painter.isOpaquePass : painter.isOpaquePass === (color[3] === 1)) {
         // Once we switch to earcut drawing we can pull most of the WebGL setup
```
Comment (lucaswoj) [inline js/render/draw_fill.js]: Please revert your changes to this file.

**Comment by mollymerp** [comment]
I added a test to the test-suite for this PR [here](https://github.com/mapbox/mapbox-gl-test-suite/pull/81)

File: debug/site.js [hunk 5]
```diff
@@ -83,12 +83,12 @@ map.on('click', function(e) {
...
+    map.tileDebug = !!this.checked;
 };
 
-document.getElementById('show-symbol-collision-boxes-checkbox').onclick = function() {
-    map.showCollisionBoxes = !!this.checked;
+document.getElementById('collision-debug-checkbox').onclick = function() {
+    map.collisionDebug = !!this.checked;
 };
 
 document.getElementById('buffer-checkbox').onclick = function() {
```
Comment (lucaswoj) [inline debug/site.js]: Please revert your changes to this file

File: js/render/draw_background.js [hunk 6]
```diff
@@ -3,7 +3,8 @@
 var TilePyramid = require('../source/tile_pyramid');
 var pyramid = new TilePyramid({ tileSize: 512 });
 var util = require('../util/util');
-var EXTENT = require('../data/bucket').EXTENT;
+var Tile = require('../source/tile');
+var pixelsToTileUnits = new Tile().pixelsToTileUnits;
```
Comment (lucaswoj) [inline js/render/draw_background.js]: We should move `pixelsToTileUnits` to a separate module (i.e.`../source/pixels_to_tile_units`). The way this is written, we have created unnecessary coupling between `draw_background` and internal implementation details of `Tile`.

**Comment by kelvinabrokwa** [comment]
@mollymerp thanks for picking this up!

File: js/render/draw_background.js (hunk 6 above)
Comment (mollymerp) [inline js/render/draw_background.js]: @lucaswoj `pixelsToTileUnits` is used in `draw_fill` `draw_line` `draw_symbol` and `painter` as well. should I split this refactor into another PR or keep it in this one?

Comment (lucaswoj) [inline js/render/draw_background.js]: We should keep those changes in this PR. They may go in a separate commit, though.

**Comment by ansis** [comment]
Looks good to me. want to merge into master?
